File: ELEC881_Tabnine/src/Function Name Masked/Python/q19-can-i-win.py
class abc(object):
    def sdf(self, maxChoosableInteger, desiredTotal):
        # Implement the solution to find the number of ways to collect all the numbers from 1 to maxChoosableInteger
        # with a total of desiredTotal
        dp = [0] * (desiredTotal + 1)
        dp[0] = 1

        for i in range(1, maxChoosableInteger + 1):
            for j in range(desiredTotal, i - 1, -1):
                dp[j] += dp[j - i]

        return dp[desiredTotal]
        

File: ELEC881_Copilot/src/No Description Prompt/Java script (JS)/q19-can-i-win.js
/**
 * @param {number} maxChoosableInteger
 * @param {number} desiredTotal
 * @return {boolean}
 */
var canIWin = function(maxChoosableInteger, desiredTotal) {
    if (desiredTotal <= maxChoosableInteger) return true;
    if ((1 + maxChoosableInteger) * maxChoosableInteger / 2 < desiredTotal) return false;
    const dp = new Map();
    const helper = (total, state) => {
        if (total <= 0) return false;
        if (dp.has(state)) return dp.get(state);
        for (let i = 1; i <= maxChoosableInteger; i++) {
            if ((state & (1 << i)) === 0) {
                if (!helper(total - i, state | (1 << i))) {
                    dp.set(state, true);
                    return true;
                }
            }
        }
        dp.set(state, false);
        return false;
    };
    return helper(desiredTotal, 0);
    
};